dali/python/nvidia/dali/plugin/tf.py

```diff
@@ -430,29 +517,104 @@ def _get_dataset(value):
                 batched = _get_external_source_param(input_name, as_input, name_es_map, 'batch')
-                input_batched_list.append(batched if batched is not None else True)
+                in_batched_list.append(batched if batched is not None else True)
+
+            return in_datasets_list, in_names_list, in_layouts_list, in_batched_list
+
+
+        def _input_lists_from_source(self, callbacked_es_map):
+
+            # TODO(klecki): Warn about this in the doc.
+            # We do it only when the users wants to use ExternalSource with `source` specified,
+            # as it has some additional limitations.
+
+            # Capture the device that DALI was placed, as we may need to copy the CPU callbacks
+            # to that device.
+            dali_device_spec = _get_current_device_spec()
+            is_dali_on_gpu = dali_device_spec.device_type == "GPU"
+
+            in_datasets_list = []
+            in_names_list = []
+            in_layouts_list = []
+            in_batched_list = []
+
+            for input_name, external_source in callbacked_es_map.items():
+                in_names_list.append(input_name)
+                layout = _get_external_source_param(input_name, None, callbacked_es_map, 'layout')
+                in_layouts_list.append(layout or "")
+
+                # Batched mode is supported by default
+                batched = _get_external_source_param(input_name, None, callbacked_es_map, 'batch')
+                in_batched_list.append(batched if batched is not None else True)
+
+                source_desc = external_source._op._source_desc
+                if source_desc.cycle == 'raise':
+                    raise NotImplementedError(("External Source node: '{}' got argument "
+                                               "cycle='raise' which is not supported.").format(input_name))
+
+                # All generator datasets must be placed on CPU.
+                with tf.device('/cpu:0'):
+                    tf_gen, dtype, shape = _get_generator_from_source_desc(
+                        source_desc, self._batch_size, external_source._batch)
+                    # dataset = tf.data.Dataset.from_generator(tf_gen, output_types=dtype)
+                    signature = _get_signature(dtype, shape)
+                    dataset = tf.data.Dataset.from_generator(tf_gen, output_signature=signature)
+                    if _cycle_enabled(source_desc.cycle):
+                        dataset = dataset.repeat()
+                    # if DALIDataset was placed on GPU, we need to add the copy targetting
+                    # that device (with proper id).
+                    if is_dali_on_gpu:
+                        dataset = dataset.apply(tf.data.experimental.copy_to_device(dali_device_spec.to_string()))
+                    in_datasets_list.append(dataset)
+
+            return in_datasets_list, in_names_list, in_layouts_list, in_batched_list
+
+
+        def _setup_inputs(self, input_datasets):
```

Comment:
Need to guard it a bit more from being used on Pipelines that don't have the inputs, so we don't have to build those pipelines.
After that I will probably be able to revert the changes in notebooks - now it introduced a breaking change it seems.